enhancements/network/ip-interface-selection.yaml

```diff
@@ -0,0 +1,302 @@
+number of requests for this.
+
+### Goals
+
+Ensure that all host networked services on a node have consistent interface
+and IP selection.
+
+### Non-Goals
+
+Support for platforms other than UPI (None).
+
+Complete support for multiple nics, with full control over what traffic
+gets routed where. However, this work should be able to serve as the basis
+for a broader multi-nic feature so we should avoid designs that would limit
+future work in this area.
+
+## Proposal
+
+The following is a possibly incomplete list of places that we do IP/interface
+selection. Ideally these should all use a single mechanism to do that so they
+are all consistent with each other.
+
+- Node IP (Kubelet and CRIO)
+- configure-ovs
+- resolv-prepender
+- Keepalived
+- Etcd
+
+In some cases (local DNS) it may not matter if the IP selected is consistent
+with the other cases, but Node IP, configure-ovs, and Keepalived all need to
+match because their functionality depends on it. I'm less familiar with the
+requirements for Etcd, but it seems likely that should match as well. In
+general, it seems best if all IP selection logic comes from one place, whether
+it's strictly required or not.
+
+### Workflow Description
+
+At deployment time the cluster administrator will include a manifest that sets
+KUBELET_NODEIP_HINT appropriately. The nodeip-configuration service (which
+will now be set as a dependency for all other services that need IP/interface
+selection) will use that value to determine the desired IP and interface for
+all services on the node. It will write the results of this selection to a
+well-known location which the other services will consume. This way, we don't
+need to duplicate the selection logic to multiple places. It will happen once
+and be reused as necessary.
+
+#### Example
+
+- resolv-prepender has to run before any other node ip selection can take
+  place. Without resolv.conf populated the nodeip-configuration service cannot
+  pull the runtimecfg image.
+  - Note: This is not relevant for plain UPI deployments, but there are some
+    deployment mechanisms that use the IPI-style resolv.conf management. Since
+    this is not present in all deployments, we will likely need to leave this
+    as it is today. The prepender script uses the runtimecfg node-ip logic,
+    so although it won't be able to consume the output of nodeip-configuration
+    there won't be any duplicate logic either.
+- nodeip-configuration runs and selects one or more IPs. It writes them to
+  the Kubelet and CRIO configuration files (this is the existing behavior).
+- nodeip-configuration also writes the following files (new behavior):
+  - /run/nodeip-configuration/primary-ip
+  - /run/nodeip-configuration/ipv4
+  - /run/nodeip-configuration/ipv6
+  - /run/nodeip-configuration/interface
+- When configure-ovs runs, it looks for the interface file written by
+  nodeip-configuration. If found, br-ex will be created on that interface. If
```

Comment:
Question: Wouldn't configure-ovs be able to rely on the fact that nodeip-configuration always chooses an interface? That way, configure-ovs also doesn't have to deal with things such as stable interfaces selection ... all of that work would be done by nodeip-configuraiton, and that always. So, the "if not" case should never happen?

Reply:
Ok but configure-ovs still needs to be able to select an interface because of IPI deployments, correct?

Reply:
After reboot nodeip-configuration will run and choose br-ex as interface but configure-ovs cleanups bridges before configuration. In that case we should not update interface file and leave old value or change configure-ovs logic

Reply:
As a first drop  i think we can use this logic only in case node-ip-hint was set. This will unblock SNO and will give sometime to get used to this new logic without breaking anything else. WDYT?

Reply:
On the other hand, in regular BM installation we will have the same problem. kubelet ip will be set according to vip but ovs script will set something else.

Reply:
I think I prefer keeping the logic in configure-ovs simpler over limiting the impact of this change. I also probably need to re-title this since, as you noted, IPI is affected even if it can't use KUBELET_NODEIP_HINT.